docs/src/rfcs/002-shader-stage-interpolants.md

```diff
@@ -0,0 +1,188 @@
+| geometry output | maxGeometryOutputComponents / 4 | 128 | 128
+| fragment input | maxFragmentInputComponents / 4 | 64 | 128
+| fragment output | maxFragmentOutputAttachments | 4 | 8
+
+
+### GLSL Linkage mechanism
+
+Historically GLSL used to link shader stage interpolants by name, however in recent years it acquired the `layout(location = N, component = K)` attribute on interpolants. GLSL also allows partial linking where a previous stage can output more data then the stage before.
+
+For pixel shader outputs in modern GLSL, instead of return a value from a function one writes to a global `out` type. And similarly, shader inputs also come from globals.
+
+### Decorators
+
+Similar to DirectX, Vulkan has the following decorations that can be applied to interpolants.
+
+| Vulkan | DirectX equivalent
+| ---- | ---- |
+| - | linear |
+| Flat| nointerpolation
+| NoPerspective| noperspective
+| Centroid | centroid
+| Sample| sample
+
+Vulkan has some additional explanations about these decorations as well:
+
+> A variable decorated with `Flat` will not be interpolated. Instead, it will have the same value for every fragment within a triangle. This value will come from a single provoking vertex. A variable decorated with `Flat` can also be decorated with Centroid or Sample, which will mean the same thing as decorating it only as `Flat`.
+
+## Metal shading language
+
+__Mostly todo.__
+
+One thing to note as a different to SPIR-V, GLSL and HLSL is that MSL doesn't seem to require explicit binding semantics or `Location` specifiers. It's specification has a `[[attribute(N)]]` specifier, but it's only used on tesselation control points and vertex shader inputs. It's specifically not used for communication between shader stages.
+
+# SPIR-V Practicalities
+
+Vertex shaders allow `Location` and `Component` decorations on input variable declarations. The `Location` decoration specifies which vertex input attribute is used to read and interpret the data that a variable will consume. The `Component` decoration allows the location to be more finely specified for scalars and vectors, down to the individual components within a location that are consumed. The components within a location are 0, 1, 2, and 3. A variable starting at component N will consume components N, N+1, N+2, …​ up through its size. For single precision types, it is invalid if the sequence of components gets larger than 3.
+
+
+## Proposal
+
+```rust
+// these `attribute`'s would match the input assembler setup on the CPU
+struct VertexInput {
+    #[attribute(0)]
+    position: Vec3,
+    #[attribute(1)]
+    normal: Vec3,
+    #[attribute(2)]
+    color: Vec3,
+}
+
+// naming this is hard, since vertex shader won't really know what shader
+// stages will run after it until the pipeline state object gets created
+// in vulkan. a tesselation shader may run after this for example,
+// and that decision won't be known by our compiler.
+struct VertexOutput {
+    #[position]
```

Comment:
Just checking, does this replace the semantics for [builtins](https://embarkstudios.github.io/rust-gpu/book/attributes.html#builtins), and all of those builtins will be supported, in both input+outputs?

Reply:
Only a relatively small amount of those builtins actually make sense in this context - for example one can't output a `VertexId` and some of those are shader-stage specific like `TessLevelOuter` or `LocalInvocationId` so I would suggest we curate a list of attributes that we allow as inputs and outputs per shader stage.